[LinearMediaPlayer] Blank video frames seen before docking and after undocking

aestes opened the pull request:
#### d5fddb160c2739f7d90e5a4b9b5ed1d6e30d69eb
<pre>
[LinearMediaPlayer] Blank video frames seen before docking and after undocking
<a href="https://bugs.webkit.org/show_bug.cgi?id=276453">https://bugs.webkit.org/show_bug.cgi?id=276453</a>
<a href="https://rdar.apple.com/130425275">rdar://130425275</a>

Reviewed by Jer Noble.

Two issues resulted in blank frames briefly appearing in an &lt;video&gt; element in element fullscreen:

1. Just prior to docking, an LMPlayableViewController is presented that obscures the web view.
Since this VC&apos;s presentationMode is immediately changed to `.fullscreenFromInline` which hides the
web view&apos;s scene and presents a new fullscreen scene, we should hide the view controller while it&apos;s
in `.inline` mode (by setting its root view&apos;s alpha to 0).

2. Right after undocking, when committing the staged AVSampleBufferDisplayLayer, MediaPlayerPrivate
would call renderingModeChanged(), which calls Element::invalidateStyleAndLayerComposition(). This
results in the &lt;video&gt; element briefly rendering a black frame. To avoid this, changed
MediaPlayerPrivate to call renderingModeChanged() as soon as the AVSampleBufferDisplayLayer is
staged, while the LinearMediaKit scene is still occluding the web view.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateDisplayLayerAndDecompressionSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayerOrVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateDisplayLayerAndDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::ensureLayerOrVideoRenderer):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):

Canonical link: <a href="https://commits.webkit.org/280848@main">https://commits.webkit.org/280848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e74ffdd9f98715f5caaf9a30aff4b2bcb5b5d5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37158 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61452 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8275 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8463 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46858 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5876 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59860 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49978 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27686 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31618 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7271 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7279 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53566 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7541 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63135 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7612 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54079 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1750 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49989 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54201 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12792 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1483 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32987 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34073 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35157 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard reviewed (commented):
I feel like this is abusing a side effect of what happens when MediaPlayer::renderingModeChanged() is called.
So this is fixing the side-effect rather than the cause.

`renderingModeChanged()` should only ever be called when the value returned by `MediaPlayerPrivate::supportsAcceleratedRendering()` changes.

Calling `renderingModeChanged()` causes a full re-layout calculation.

One issue currently is MediaPlayerPrivateRemote::supportsAcceleratedRendering currently returns false by default until the MediaPlayerPrivate is created in the GPU process and has finished initialising.

So our implementation incorrectly used of when the AVSampleBufferDisplayLayer was created to determine when `supportsAcceleratedRendering()` will return true. And that is an incorrect way of doing it.

The MediaPlayerPrivateMediaSourceAVFObjC::supportsAcceleratedRendering() always returns true and same with MediaPlayerPrivateAVFoundationObjC::supportsAcceleratedRendering()

So they really should never call `renderingModeChanged()` themselves ; but the `RemoteMediaPlayerProxy` should instead make the call as soon as the MediaPlayerPrivate is created in the GPU process.

I had avoided making this change in the current release for the fear of breaking things as this code has proven itself to be rather fragile and the consequences has always been about having a black video playback.

Maybe we should do it properly once and for all now

aestes committed:
[LinearMediaPlayer] Blank video frames seen before docking and after undocking
https://bugs.webkit.org/show_bug.cgi?id=276453
rdar://130425275

Reviewed by Jer Noble.

Two issues resulted in blank frames briefly appearing in an <video> element in element fullscreen:

1. Just prior to docking, an LMPlayableViewController is presented that obscures the web view.
Since this VC's presentationMode is immediately changed to `.fullscreenFromInline` which hides the
web view's scene and presents a new fullscreen scene, we should hide the view controller while it's
in `.inline` mode (by setting its root view's alpha to 0).

2. Right after undocking, when committing the staged AVSampleBufferDisplayLayer, MediaPlayerPrivate
would call renderingModeChanged(), which calls Element::invalidateStyleAndLayerComposition(). This
results in the <video> element briefly rendering a black frame. To avoid this, changed
MediaPlayerPrivate to call renderingModeChanged() as soon as the AVSampleBufferDisplayLayer is
staged, while the LinearMediaKit scene is still occluding the web view.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateDisplayLayerAndDecompressionSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayerOrVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateDisplayLayerAndDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::ensureLayerOrVideoRenderer):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):

Canonical link: https://commits.webkit.org/280848@main
